### ISSUE TITLE
New lint: detect homoglyphs

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust-clippy#2368]
Homoglyphs are different unicode characters that to the naked eye look the same.

```rust
#![feature(non_ascii_idents)]
fn main() {
    let a = 2;
    let а = 3;
    assert_eq!(a, 2);
    assert_eq!(а, 3);
}
```

where `a` is latin lower case a, and `а` is cyrillic lower case a. Another example is: A, Α, А (Latin-A, Greek Alpha, Cyrillic-A).

This binary runs on my machine without errors, but it seems that the Rust playground doesn't support unicode yet.

Homoglyph-based attacks are typically used in domain names, and underhanded code.

To prevent this attacks we can check whether any two identifiers in scope are homoglyphs, and warn about it.

**Comment by oli-obk** [comment]
Rustc has a list for symbols already: https://github.com/wafflespeanut/rust/blob/7f63c7cf4c2938c31de3d63fc769706f0d87cb54/src/libsyntax/parse/lexer/unicode_chars.rs

The full list of confusable unicodes is http://www.unicode.org/Public/security/revision-06/confusables.txt

We should probably automatically parse that file into an array and use that (instead of attempting any manual transcription/processing of the file)

**Comment by Centril** [comment]
cc https://github.com/rust-lang/rust/issues/55467.

**Comment by phansch** [comment]
Since it seems like it's going to be implemented in rustc directly, I'm going to go ahead and close this issue.

The tracking issue linked above is still relevant. There already is a `non_ascii_identifiers` lint and a [PR was opened](https://github.com/rust-lang/rust/pull/71542) to implement homoglyph detection for identifiers as well.
